driver-core/src/test/java/com/datastax/driver/core/CustomPayloadTest.java

```diff
@@ -0,0 +1,263 @@
+        assertPayloadsEqual(payload1, actual);
+    }
+
+    // prepare
+
+    /**
+     * Ensures that an incoming payload is propagated from prepared to bound statements.
+     * @throws Exception
+     */
+    @Test(groups = "short")
+    public void should_propagate_incoming_payload_to_bound_statement() throws Exception {
+        RegularStatement statement = new SimpleStatement("SELECT c1, c2 FROM t1 where c1 = ?");
+        statement.setOutgoingPayload(payload1);
+        PreparedStatement ps = session.prepare(statement);
+        assertPayloadsEqual(payload1, ps.getIncomingPayload());
+        BoundStatement bs = ps.bind(1);
+        ResultSet rows = session.execute(bs);
+        Map<String,ByteBuffer> actual = rows.getExecutionInfo().getIncomingPayload();
+        assertPayloadsEqual(payload1, actual);
+        bs = ps.bind();
+        bs.setInt(0, 1);
+        rows = session.execute(bs);
+        actual = rows.getExecutionInfo().getIncomingPayload();
+        assertPayloadsEqual(payload1, actual);
+    }
+
+    /**
+     * Ensures that an incoming payload is overridden by an explicitly set outgoing payload
+     * when propagated to bound statements.
+     * @throws Exception
+     */
+    @Test(groups = "short")
+    public void should_override_incoming_payload_when_outgoing_payload_explicitly_set_on_preparing_statement() throws Exception {
+        RegularStatement statement = new SimpleStatement("SELECT c2, c1 FROM t1 where c1 = ?");
+        statement.setOutgoingPayload(payload1);
+        PreparedStatement ps = session.prepare(statement);
+        ps.setOutgoingPayload(payload2);
+        assertPayloadsEqual(payload1, ps.getIncomingPayload());
+        BoundStatement bs = ps.bind(1);
+        ResultSet rows = session.execute(bs);
+        Map<String,ByteBuffer> actual = rows.getExecutionInfo().getIncomingPayload();
+        assertPayloadsEqual(payload2, actual);
+        bs = ps.bind();
+        bs.setInt(0, 1);
+        rows = session.execute(bs);
+        actual = rows.getExecutionInfo().getIncomingPayload();
+        assertPayloadsEqual(payload2, actual);
+    }
+
+    // pagination
+
+    /**
+     * Ensures that a custom payload is propagated throughout pages.
+     * @throws Exception
+     */
+    @Test(groups = "short")
+    public void should_echo_custom_payload_when_paginating() throws Exception {
+        session.execute("INSERT INTO t1 (c1) VALUES (1)");
+        session.execute("INSERT INTO t1 (c1) VALUES (2)");
+        Statement statement = new SimpleStatement("SELECT c2 FROM t1 where c1 IN (1,2)");
+        statement.setFetchSize(1);
+        statement.setOutgoingPayload(payload1);
+        ResultSet rows = session.execute(statement);
+        rows.all();
+        for (ExecutionInfo info : rows.getAllExecutionInfo()) {
+            Map<String,ByteBuffer> actual = info.getIncomingPayload();
+            assertPayloadsEqual(payload1, actual);
+        }
+    }
+
+    // TODO retries, spec execs
```

Comment:
Until then we have a repeatable way of testing those cases (i.e. Scassandra with proto V4) I'll make sure to try this out manually to make sure everything is kosher.